common/src/main/java/io/netty/util/internal/BoundedInputStream.java

```diff
@@ -37,47 +37,37 @@ public BoundedInputStream(@NotNull InputStream in) {
 
     @Override
     public int read() throws IOException {
-        checkMaxBytesRead(1);
-        try {
-            int b = super.read();
-            if (b <= 0) {
-                // We couldn't read anything.
-                numRead--;
-            }
-            return b;
-        } catch (IOException e) {
-            numRead--;
-            throw e;
+        checkMaxBytesRead();
+
+        int b = super.read();
+        if (b > 0) {
+            numRead++;
         }
+
+        checkMaxBytesRead();
+        return b;
     }
 
     @Override
     public int read(byte[] buf, int off, int len) throws IOException {
+        checkMaxBytesRead();
+
         // Calculate the maximum number of bytes that we should try to read.
         int num = Math.min(len, maxBytesRead - numRead + 1);
-        checkMaxBytesRead(num);
```

Comment:
The original motivation for checking for `num` was to fail before starting the actual read (no need to try reading if we know we are going out of bound). Is it possible to preserve that behavior?

Reply:
`maxBytesRead` already protects from "don't read more than allowed", but it is worth reading if the requested `len` is larger than `maxBytesRead`?

Reply:
> The original motivation for checking for `num` was to fail before starting the actual read (no need to try reading if we know we are going out of bound). Is it possible to preserve that behavior?

No, it is not possible to do that.  You have no idea when the underlying stream will terminate, so just because the request is to read `len` bytes, you cannot assume that many bytes will actually be read.

Reply:
Ok, so for the cases when the caller doesn't know about the underlying file size and the `BoundedInputStream` protection, it can still request more which doesn't mean we read that much. Makes sense to me, thanks for clarifying.